config-svc/pkg/api/routes_v1.go

```diff
@@ -15,93 +15,137 @@
 package api
 
 import (
-	"encoding/json"
-	"errors"
 	"fmt"
-	"io/fs"
+	"io"
 	"net/http"
 	"os"
 
-	"github.com/kennygrant/sanitize"
+	"github.com/couchbase/tools-common/cbvalue"
+	"github.com/couchbaselabs/observability/config-svc/pkg/couchbase"
+	"github.com/couchbaselabs/observability/config-svc/pkg/prometheus"
+	"gopkg.in/yaml.v3"
 
 	v1 "github.com/couchbaselabs/observability/config-svc/pkg/api/v1"
 	"github.com/labstack/echo/v4"
 )
 
-const (
-	prometheusTargetsPathDevelopment = "./targets/%s.json"
-	prometheusTargetsPathProduction  = "/etc/prometheus/couchbase/custom/%s.json"
-)
-
-type prometheusTargets []struct {
-	Targets []string          `json:"targets"`
-	Labels  map[string]string `json:"labels"`
-}
+const defaultPrometheusConfigPath = "/etc/prometheus/prometheus.yml"
 
-func (s *Server) PostAddPrometheusTarget(ctx echo.Context) error {
-	var data v1.PostAddPrometheusTargetJSONBody
+func (s *Server) PostClustersAdd(ctx echo.Context) error {
+	var data v1.PostClustersAddJSONRequestBody
 	if err := ctx.Bind(&data); err != nil {
 		return err
 	}
 
-	labels := data.Labels.AdditionalProperties
-	if labels == nil {
-		labels = make(map[string]string)
+	secure := false
+	if data.CouchbaseConfig.Secure != nil && *data.CouchbaseConfig.Secure {
+		secure = true
 	}
-	if data.NameLabel != nil {
-		labels[*data.NameLabel] = data.Name
-	}
-	value := prometheusTargets{
-		{
-			Targets: data.Targets,
-			Labels:  labels,
-		},
+	mgmtPort := 8091
+	if data.CouchbaseConfig.ManagementPort != nil {
+		mgmtPort = int(*data.CouchbaseConfig.ManagementPort)
 	}
-	if value[0].Labels == nil {
-		value[0].Labels = make(map[string]string)
-	}
-	valueJSON, err := json.Marshal(value)
+	cluster, err := couchbase.FetchCouchbaseClusterInfo(
+		data.Hostname,
+		mgmtPort,
+		secure,
+		data.CouchbaseConfig.Username,
+		data.CouchbaseConfig.Password,
+	)
 	if err != nil {
 		return err
 	}
 
-	var path string
-	if s.production {
-		path = prometheusTargetsPathProduction
-	} else {
-		path = prometheusTargetsPathDevelopment
+	staticConfig := prometheus.StaticConfig{
+		Targets: make([]string, len(cluster.Nodes)),
+		Labels: map[string]string{
+			"cluster": cluster.ClusterName,
+		},
 	}
-	path = fmt.Sprintf(path, sanitize.Name(data.Name))
 
-	if data.Overwrite == nil || !*data.Overwrite {
-		_, err := os.Stat(path)
-		if err == nil {
-			return echo.NewHTTPError(http.StatusConflict, "target already exists")
-		} else if errors.Is(err, fs.ErrNotExist) {
-			// Ignore it
-		} else {
+	var anyNodeCB7 bool
+
+	for i, node := range cluster.Nodes {
+		hostname, mgmtPort, err := node.ResolveHostPort(secure)
+		if err != nil {
 			return err
 		}
+		if node.Version.AtLeast(cbvalue.Version7_0_0) {
+			staticConfig.Targets[i] = fmt.Sprintf("%s:%d", hostname, mgmtPort)
+			anyNodeCB7 = true
```

Comment:
It might be better just to have a variable for the authentication stuff you add if it is CBS 7, currently all this flag does is later on be used for that so why not construct here. You also do not need to re-check it as soon as you've found one.

I'm not sure how this works when basic auth is set for the exporter though?

Reply:
We'd still need to re-check it, for the (rare yet not impossible) case of a mixed 6.x and 7.x cluster, where some would need the management port and others would need the exporter port.

Reply:
Ah yeah I see what you mean.

Do you handle basic auth for the exporter though?

Reply:
Yes (*)

(*) implicitly, by assuming the credentials are the same for CBS, if they're different then no

Reply:
Probably need to capture that then but low priority.